torch/csrc/jit/codegen/cuda/lower2device.cpp

```diff
@@ -19,22 +20,163 @@
 namespace cuda {
 
 // TODO(kir): revisit this
 thread_local GpuLower* active_gpu_lower = nullptr; // NOLINT
+namespace {
+
+struct idcomp {
+  bool operator()(const IterDomain* lhs, const IterDomain* rhs) const {
+    return lhs->name() < rhs->name();
+  }
+};
+
+// Going to generate a map of tensor view root domain extents to reduce the
+// number used during lowering. For example if we have:
+//
+// T2[i0, i1] = T1[i0, i1] + T2[i2, i3]
+//
+// We know it would be safe to use:
+//
+// T2[i0, i1] = T1[i0, i1] + T2[i0, i1]
+//
+// And that way we don't generate T2.size[0] and T2.size[1], instead we will
+// reuse T1.size[0] and T1.size[1]
+// This is important when doing CSE as T2 and T1 would otherwise look like
+// they're using different values, even though we know they're the same
+//
+// There's some duplicate logic here that's in computeAt map, but it's not so
+// concice there to pull out. May want to consider making this mapping its own
+// class especially as it may be useful during scheduling.
+std::unordered_map<Val*, Val*> getSimplificationMap(Fusion* fusion) {
+  std::list<std::unordered_set<IterDomain*>> disjoint_root_sets;
+  std::unordered_map<IterDomain*, std::unordered_set<IterDomain*>*>
+      id_to_disjoint_root_set;
+
+  auto map_root_ids = [&disjoint_root_sets, &id_to_disjoint_root_set](
+                          IterDomain* id0, IterDomain* id1) {
+    if (id0->isBroadcast() || id1->isBroadcast()) {
+      return;
+    }
+
+    auto disjoint_set_0_it = id_to_disjoint_root_set.find(id0);
+    auto disjoint_set_1_it = id_to_disjoint_root_set.find(id1);
+    bool set_0_found = disjoint_set_0_it != id_to_disjoint_root_set.end();
+    bool set_1_found = disjoint_set_1_it != id_to_disjoint_root_set.end();
+
+    if (set_0_found && set_1_found) {
+      if (disjoint_set_0_it->second == disjoint_set_1_it->second) {
+        return;
+      }
+      // merge second disjoint set into first
+      auto* set_0 = disjoint_set_0_it->second;
+      auto* set_1 = disjoint_set_1_it->second;
+      for (auto id : *set_1) {
+        set_0->emplace(id);
+        id_to_disjoint_root_set[id] = set_0;
+      }
+      // remove second set from disjoint_root_sets
+      disjoint_root_sets.erase(std::find(
+          disjoint_root_sets.begin(), disjoint_root_sets.end(), *set_1));
+    } else if (set_0_found || set_1_found) {
+      auto existing_set =
+          set_0_found ? disjoint_set_0_it->second : disjoint_set_1_it->second;
+      auto to_add_id = set_0_found ? id1 : id0;
+      existing_set->emplace(to_add_id);
+      id_to_disjoint_root_set[to_add_id] = existing_set;
+      // add entry into existing set
+    } else {
+      // create new set entry
+      disjoint_root_sets.push_back(std::unordered_set<IterDomain*>());
+      auto* new_set = &disjoint_root_sets.back();
+      new_set->emplace(id0);
+      new_set->emplace(id1);
+      id_to_disjoint_root_set[id0] = new_set;
+      id_to_disjoint_root_set[id1] = new_set;
+    }
+  };
```

Comment:
There's `DisjointSet` originally written by @shmsong. Would it be usable here to simplify the implementation?

Reply:
Thanks, forgot it still existed. Will try to convert over to it.